python/lsst/daf/butler/core/s3utils.py

```diff
@@ -19,19 +19,30 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-__all__ = ("s3CheckFileExists", "bucketExists", "setAwsEnvCredentials",
+__all__ = ("getClient", "s3CheckFileExists", "bucketExists", "setAwsEnvCredentials",
            "unsetAwsEnvCredentials")
 
 import os
-
-try:
-    import boto3
-except ImportError:
-    boto3 = None
+import boto3
 
 from .location import ButlerURI, Location
 
 
+def getClient():
+    """Return a s3 client
+
+    Look for the crendentials in the env, and connect to the S3 client
+    with the default AWS or the Google Cloud Storage endpoint
+    """
+    try:
+        key = os.environ['AWS_ACCESS_KEY_ID']
```

Comment:
I very much disagree with this choice. This completely locks us into only 1 way of authentication - and the worst and most unsafe one at that. If you want to lock GCE into this authentication strategy I guess that's fine by me but I instantiate a generic client to take advantage of the different, much safer, authentication mechanisms such as IAM, `~/.aws etc... 

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#configuring-credentials

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#best-practices-for-configuring-credentials

Reply:
I'm happy to take your input on whether the use of the environment variable is the right thing. Do you think that getS3Client needs to exist at all? Or would configuration changes mean that no code changes are needed at all here?

Reply:
Hm, the problem is that I don't know how GCE handles authentication tbh. From the code it seems to me that HFC needed to pass in the endpoint URL and that it didn't work out for her otherwise. 
I don't want to have two different classes for object datastore if the only thing that's different is that they instantiate the client differently, but if boto3 needs the enpoint URL that seems like the way to go to me. I guess the other option is to wrap the `boto3.client('s3')`, catch the authentication exception, try wih google endpoint instead and reraise if required Then everything can remain the same.

I assume that another benefit to having two classes is that then we there is enough flexibility to diverge and adjust to individual object store specifics. But again, I don't know if there are any such use-cases so many not run before I walk (i.e. I need to do some reading up on it).

Reply:
I'm not sure I understand why we would ever want to have different AWS and GCE datastore implementations if the only difference is the endpoint URL. We clearly need a way to specify whether we are using AWS, Google, or US Data Facility S3 endpoint. This has to happen presumably before you try to access a butler.yaml file from using s3:// URI so can't be in that butler.yaml file (it's too late by then).  This patch tries to put all the code that could use an endpoint_url in a single place and then the next question is how that code can work out where to look. We either use environment variables or we use some other dot file. This is orthogonal to credentials.

Reply:
Also, always calling out to AWS, failing, and then trying Google, does not seem like the right answer to me.

Reply:
Doing this doesn't prohibit the authentication mechanism through `~/.aws/` for AWS S3.  So for AWS things should continue to work.   Besides `~/.aws` are there other methods I can test? 

It does mean for GCS we would only use the env var way for now. There are more ways of authenticating with GCS but I haven't looked much.  I can create a ticket to investigate more and add more authentication methods in later.

Reply:
And presumably one option to consider is a DAF_BUTLER_S3_ENDPOINT environment variable that will contain the relevant endpoint URL. Then the code just reads that directly from the environment and doesn't have to guess about anything (with unset being None being AWS).

Reply:
@r-owen how are T&S specifying the endpoint URL?

Reply:
ts_salobj uses env variable S3_ENDPOINT_URL as per https://ts-salobj.lsst.io/configuration.html#environment-variables

Reply:
Ok, I'm wrong. Sorry about that. I guess I was in too much of a rush to give it a good thought. 

I still see many benefits to having a specialized datastore for GCS using a GCS SDK instead of the [Interop GCS API](https://cloud.google.com/storage/docs/interoperability). For example, as mentioned in the docs, any `boto3` functionality calling `list_objects_v2` (which is the [new and recommended](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjects.html) way to do it) under the hood wouldn't work, it's not supported. I assume this was the reason the Buckets API in tests had to be replaced. 

I still think allowing only strictly environmental variable authentication is not a good idea. We are stuffing Google Cloud Storage keys under `AWS_ACCESS_KEY_ID` values. I agree with Tim's latest suggestion that we should at least rename the env var name.

Reply:
Ok. Now I understand your other comment. We are talking about multiple things in parallel. The first is "how do we specify arbitrary S3 endpoints" and the second is "in the future it might be worth writing a Datastore that uses google-native APIs rather than S3 compatibility layer".   I think we have agreed that an endpoint environment variable specifying the actual endpoint to use is a reasonable solution. I am more than happy to end up with a google-specific datastore at some point in the future.

Reply:
We are also talking about two kinds of env var here.  One is the endpoint, and the other is the authentication key.

For the authentication key, for GCS using env var `AWS_ACCESS_KEY_ID` is the quickest way to migrate to GCS but we will want to look into other ways too.  For AWS the authentication mechanism is not changed; AWS credential files continue to work like before.    

For the endpoint it sounds like we want to read it from the env var `S3_ENDPOINT_URL` like ts_salobj does, instead of coding it in.  My current patch is to use `AWS_ACCESS_KEY_ID` to know if it's GCS or AWS, so avoiding one more env var.

Reply:
The patch has been replaced by one using the env var `S3_ENDPOINT_URL`. It should be less confusing now.

DM-25093 is created for looking into other GCS authentication methods.